src/main/java/io/aiven/kafka/connect/s3/S3SinkTask.java

```diff
@@ -46,40 +74,159 @@ public S3SinkTask() {
+                    )
+                );
+        if (Objects.isNull(awsEndpointConfig)) {
+            s3ClientBuilder.withRegion(config.getAwsS3Region());
+        } else {
+            s3ClientBuilder.withEndpointConfiguration(awsEndpointConfig).withPathStyleAccessEnabled(true);
+        }
+        s3Client = s3ClientBuilder.build();
+        outputWriter = getOutputWriter();
+
+        try {
+            recordGrouper = RecordGrouperFactory.newRecordGrouper(config);
+        } catch (final Exception e) {
+            throw new ConnectException("Unsupported file name template " + config.getFilename(), e);
+        }
+    }
+
+    private OutputWriter getOutputWriter() {
+        switch (this.config.getFormatType()) {
+            case CSV:
+                return OutputPlainWriter.builder().addFields(config.getOutputFields()).build();
+            case JSONL:
+                return JsonLinesOutputWriter.builder().addFields(config.getOutputFields()).build();
+            default:
+                throw new ConnectException("Unsupported format type " + config.getFormatType());
+        }
     }
 
     @Override
     public void put(final Collection<SinkRecord> records) throws ConnectException {
+        Objects.requireNonNull(records, "records cannot be null");
         LOGGER.info("Processing {} records", records.size());
-        records.stream()
-            .map(r -> Pair.of(new TopicPartition(r.topic(), r.kafkaPartition()), r))
-            .forEach(p -> streamMap.write(p.getLeft(), p.getRight()));
+        records.forEach(recordGrouper :: put);
     }
 
     @Override
     public void flush(final Map<TopicPartition, OffsetAndMetadata> offsets) {
-        offsets.keySet().forEach(streamMap::flush);
+        recordGrouper.records().forEach(this::flushFile);
+        recordGrouper.clear();
     }
 
-    @Override
-    public void close(final Collection<TopicPartition> partitions) throws ConnectException {
-        partitions.forEach(tp -> {
-            LOGGER.info("Close {}", tp);
-            streamMap.close(tp);
-        });
+    private void flushFile(final String filename, final List<SinkRecord> records) {
+        Objects.requireNonNull(records, "records cannot be null");
+        if (records.isEmpty()) {
+            return;
+        }
+        final SinkRecord sinkRecord = records.get(0);
+        try (final OutputStream compressedStream = newStreamFor(filename, sinkRecord)) {
+            for (int i = 0; i < records.size() - 1; i++) {
+                outputWriter.writeRecord(records.get(i), compressedStream);
+                compressedStream.flush();
+            }
+            outputWriter.writeLastRecord(records.get(records.size() - 1), compressedStream);
+        } catch (final IOException e) {
+            throw new ConnectException(e);
+        }
     }
 
     @Override
     public void stop() {
-        LOGGER.info("Stop  S3 Sink Task");
-        streamMap.closeAll();
+        LOGGER.info("Stop S3 Sink Task");
     }
```

Comment:
Better to call `s3Client.shutdown()` explicitly here, since S3 client runs HTTP async with threadpool

Reply:
Ok.
- [ ] Shutdown explicitly

Reply:
Yep, this seems to be missing before.